processing/src/main/java/org/apache/druid/segment/join/table/IndexedTableJoinMatcher.java

```diff
@@ -240,13 +243,36 @@ private void advanceCurrentRow()
    */
   private static class ConditionMatcherFactory implements ColumnProcessorFactory<Supplier<IntIterator>>
   {
+    private static final int MAX_NUM_CACHE = 10;
+    private static final int CACHE_MAX_SIZE = 1000;
+
     private final ValueType keyType;
     private final IndexedTable.Index index;
 
+    // DimensionSelector -> (int) dimension id -> (IntList) row numbers
+    private final LoadingCache<DimensionSelector, LoadingCache<Integer, IntList>> dimensionCaches;
+
     ConditionMatcherFactory(ValueType keyType, IndexedTable.Index index)
     {
       this.keyType = keyType;
       this.index = index;
+
+      this.dimensionCaches =
+          Caffeine.newBuilder()
+                  .maximumSize(MAX_NUM_CACHE)
+                  .build(
+                      selector ->
+                          Caffeine.newBuilder()
+                                  .maximumSize(CACHE_MAX_SIZE)
+                                  .build(dimensionId -> getRowNumbers(selector, dimensionId))
+                  );
+
+    }
+
+    private IntList getRowNumbers(DimensionSelector selector, int dimensionId)
+    {
+      final String key = selector.lookupName(dimensionId);
```

Comment:
For each selector, `dimensionId`s are contiguous and go from `0` (inclusive) to `selector.getValueCardinality()` (exclusive). If the cardinality is less than `CACHE_MAX_SIZE` then you could use an `IntList[]` as a cache. It should be faster.

Reply:
(Assuming the dictionary is real, i.e., `selector.getValueCardinality() != DimensionSelector.CARDINALITY_UNKNOWN`)